history: add "In" and "To" fields + add address label

rating89us opened the pull request:
Changes:
- removed trailing zeros in Fee

Changes in Received transations:
- added "In" field to the Blockheight field position. This field will display the receiving address index + label (from the receive page) or the truncated address (when the address has no label) (fixes #2987)
- Blockheight field moved to the bottom.
- increased height of collapsed received transactions to the same as sent transactions (320 pixels), in order to display the blockheight. As a side effect, transaction key is also being displayed for received transactions.

Changes in Sent transactions:
- moved "To" field to the Blockheight field position. This field will now display either an address book icon + a destination address label (from the address book) or a truncated destination address (when the destination address has no address book entry).
- Blockheight field moved to the bottom.
- changed "Waiting for transaction to leave txpool" message to "Waiting confirmation..." in order to fit in the smaller space and because it's easier to understand.
- transactions with amount = 0 XMR will display destination as "My wallet", since they are seen in self transactions after a wallet is rescanned or restored from keys/seed.
- in restored wallets, transactions sending to the same wallet will display "Unknown amount", since we can't recover the amount information from the blockchain

![image](https://user-images.githubusercontent.com/45968869/87012920-2c5bf500-c1ca-11ea-94d7-3387eea1061e.png)

![image](https://user-images.githubusercontent.com/45968869/87012972-3f6ec500-c1ca-11ea-958e-704dddf398f3.png)

xiphon commented:
Please split into separate PRs

rating89us commented:
> Please split into separate PRs

Done

selsta commented:
<img width="658" alt="Screenshot 2020-09-29 at 03 34 01" src="https://user-images.githubusercontent.com/7697454/94502516-9fdb3480-0204-11eb-8e8f-c3437e3d47fd.png">

Would like a bit more margin between In | Date

selsta commented:
Looks good now IMO, please squash the commits.

selsta commented on pages/History.qml:
```diff
@@ -769,14 +760,44 @@ Rectangle {
+                                                    return FontAwesome.addressBook + " " + addressBookName;
+                                                }
+                                            }
+                                            if (blockheight === 0) {
```
return (blockheight ? qsTr("Unknown recipient") : qsTr("Waiting confirmation...")) + translationManager.emptyString;

rating89us replied:
I agree

selsta commented on pages/History.qml:
```diff
@@ -769,14 +760,44 @@ Rectangle {
+                                    font.pixelSize: 15
+                                    text: {
+                                        if (isout) {
+                                            if (address !== "") {
```
minor but just `if (address)` seems a bit easier to read

inside the if you could also use the ternary operator

rating89us replied:
i agree

selsta commented on pages/History.qml:
```diff
@@ -769,14 +760,44 @@ Rectangle {
+                                        } else {
+                                            const receivingAddress = currentWallet ? currentWallet.address(subaddrAccount, subaddrIndex) : null;
+                                            const receivingAddressLabel = currentWallet ? appWindow.currentWallet.getSubaddressLabel(subaddrAccount, subaddrIndex) : null;
+                                            if (receivingAddress) {
```
else missing, resulting in

```
2020-09-29 21:15:35.309	W qrc:/pages/History.qml:764:37: Unable to assign [undefined] to QString
```

rating89us replied:
fixed

rating89us commented:
Added a new thing:
If you rescan your wallet (change wallet restore height) or restore it from keys/seed, your wallet will lose details from past transactions and all transactions that were sent to your own wallet will display an amount 0.000000000000 XMR. 

So in transactions with amount = zero, "To:" field will now display "My wallet"

![image](https://user-images.githubusercontent.com/45968869/94628647-cf4f7700-02c0-11eb-8763-d6e878d2df61.png)

selsta commented:
Unknown recipient -> You can click on it, it turns orange and it says "copied to clipboard" but nothing is in clipboard.

rating89us commented:
Fixed. It will now copy the text "My wallet" or "Unknown recipient".

Even in "My wallet" we don't know the exact receiving address, because it can be either the main address or a subaddress.

xiphon commented on pages/History.qml:
```diff
@@ -633,17 +633,7 @@ Rectangle {
-                                        }
-                                        return addressBookName;
-                                    }
+                                    text: isout ? qsTr("Sent") : qsTr("Received") + translationManager.emptyString
```
`1 ? 2 : 3 + 4` == `2`

rating89us replied:
i don't see what's wrong here

selsta replied:
add () around the ternary operator so that + emptyString gets applied to both

rating89us replied:
+ emptyString is already being applied to both Sent and Received

selsta replied:
no, it only gets applied to Received due to + having precedence over the ternary operator

rating89us replied:
in my wallet the Sent string is being translated normally. And I've already seen this type of `A ? B : C + translationManager.emptyString` many times in monero GUI code, always working fine.

selsta replied:
> in my wallet the Sent string is being translated normally.

Yes because `translationManager.emptyString` is technically not needed when using the ternary operator, not because the () are not needed.

> And I've already seen this type

Then it is wrong in these places too, would appreciate it if you could update it here, else it is wrong code that works by accident.

rating89us replied:
My code is:
`isout ? qsTr("Sent") : qsTr("Received") + translationManager.emptyString`

Actually the computer reads this (but somehow translation is working):
`(isout ? qsTr("Sent") : qsTr("Received") + translationManager.emptyString)`

And you want me to change to this:
`(isout ? qsTr("Sent") : qsTr("Received")) + translationManager.emptyString`

Right?

selsta replied:
> but somehow translation is working

Due to technically `translationManager.emptyString` not being needed at all in this case. But it is better if we still add it.

> And you want me to change to this:

Yes.

rating89us replied:
Why is translationManager.emptyString not needed in this case? I thought that I should add `+ translationManager.emptyString` at the end of every line that contains a `qsTr("");`, so that the string gets added to the translation.

selsta replied:
translationManager.emptyString is a workaround for a Qt bug, AFAIK it is only needed in cases like this:

```
text: amount
```

We add it to every string just to be sure.

xiphon commented on pages/History.qml:
```diff
@@ -768,15 +759,44 @@ Rectangle {
+                                            const receivingAddressLabel = currentWallet ? appWindow.currentWallet.getSubaddressLabel(subaddrAccount, subaddrIndex) : null;
+                                            if (receivingAddress) {
+                                                if (subaddrIndex == 0) {
+                                                    return qsTr("Address #0") + " (" + qsTr("Primary address") + ")" + translationManager.emptyString;
```
IMO you'd better move " #0" out of the translatable string

rating89us replied:
fixed

rating89us commented:
Some changes: 
1) I changed the 0.000000000000 XMR to "Unknown amount" in transactions sent to the same wallet:
- The amount entered by the user is always >0 XMR, and we can't recover it from the blockchain.
- The net balance of the transacion = the transaction fee, therefore it's negative. Displaying 0 XMR seems that the net balance = 0 XMR.

2) Removed trailing zeros from Fee

3) Added/improved details in transaction details popup

rating89us commented:
Currently the To field can only display:
- To: _address string_ (this address can be from my wallet or external wallet)
- To: "My wallet" (transaction with amount 0.00000000 XMR in a recovered wallet)
- To: "Unknown recipient" (transaction with amount different from 0.00000 XMR in a recovered wallet)

To better differentiate case 1, we need a new function to detect if a given address is from the own wallet. The input would be the address string and the function would return true if the address is from the wallet or false if it's from an external wallet.

The resulting To field possibilities would be:
- To: "My wallet (_address string_)"
- To: "External address (_address string_)"
- To: "My wallet" (transaction with amount 0.00000000 XMR in a recovered wallet)
- To: "Unknown recipient" (transaction with amount different from 0.00000 XMR in a recovered wallet)

xiphon commented on pages/History.qml:
```diff
@@ -768,15 +759,44 @@ Rectangle {
+                                                const addressBookName = currentWallet ? currentWallet.addressBook.getDescription(address) : null;
+                                                return (addressBookName ? FontAwesome.addressBook + " " + addressBookName : TxUtils.addressTruncate(address, 8));
+                                            }
+                                            if (amount != 0.000000000000) {
```
Could we use `0` instead of `0.000000000000`?

rating89us replied:
I tried, but it doesn't seem to work

selsta replied:
are you sure? seems to work here

rating89us replied:
fixed (i was wrong)

xiphon commented on pages/History.qml:
```diff
@@ -768,15 +759,44 @@ Rectangle {
+                                            const receivingAddress = currentWallet ? currentWallet.address(subaddrAccount, subaddrIndex) : null;
+                                            const receivingAddressLabel = currentWallet ? appWindow.currentWallet.getSubaddressLabel(subaddrAccount, subaddrIndex) : null;
+                                            if (receivingAddress) {
+                                                if (subaddrIndex == 0) {
+                                                    return qsTr("Address") + " #0" + " (" + qsTr("Primary address") + ")" + translationManager.emptyString;
+                                                } else {
+                                                    if (receivingAddressLabel) {
+                                                        return qsTr("Address") + " #" + subaddrIndex + " (" + receivingAddressLabel + ")" + translationManager.emptyString;
+                                                    } else {
+                                                        return qsTr("Address") + " #" + subaddrIndex + " (" + TxUtils.addressTruncate(receivingAddress, 4) + ")" + translationManager.emptyString;
+                                                    }
+                                                }
```
Account number / label is ignored

rating89us replied:
Transactions page only display transactions from the currently selected account

selsta replied:
Yea, I think account stuff is not needed here, also there is not enough space for it.

xiphon replied:
Gotcha

xiphon commented on pages/History.qml:
```diff
@@ -1510,7 +1527,7 @@ Rectangle {
                 "i": i,
                 "isout": isout,
-                "amount": Number(amount),
+                "amount": Utils.removeTrailingZeros(displayAmount.toFixed(12)),
```
Why this change is needed?

rating89us replied:
I removed Number() because amounts like 0.000001 were returning 0.

xiphon replied:
> I removed Number() because amounts like 0.000001 were returning 0.

How can i reproduce the issue?

rating89us replied:
In this example, transactions #1 to #6 the wallet knows the amount. Transaction #7 the wallet doesn't know the amount (wallet was restored at this point).

`"amount": Utils.removeTrailingZeros(displayAmount.toFixed(12))`: all transactions with known amount  (#1-6) display the amount correctly. 
![image](https://user-images.githubusercontent.com/45968869/94722985-5d2a7100-0358-11eb-9a5a-4f83910a8bf9.png)


` "amount": Number(amount)`: some transactions (#3-6) with known amount display as Unknown amount. 
![image](https://user-images.githubusercontent.com/45968869/94722858-2eac9600-0358-11eb-8df9-b7e3c10d77dc.png)

xiphon replied:
Can't reproduce. Works well with `"amount": Number(amount)` (Qt 5.9.7). What Qt version do you use?

xiphon replied:
Did you send the txes in question to the same wallet but to the different account or did you send them to a third party wallet?

selsta replied:
I checked with Qt 5.12.9 and 5.15.1 and also can’t reproduce this issue. @rating89us can you double check?

rating89us replied:
My wallet main address (account #1) is 48rg...
The first two transactions are from account #1 (48rg...) to account #2's primary address (89HP...) and its first subaddress (88PS...).
The remaining transactions are from account #1 to account #1 (48rg... to 48rg...)

I'm using Qt 5.12.9.
I'll check again if I can identify what's causing this.

rating89us replied:
There is a comment in line 1510 :
```
// *sometimes* amount is 0, while the 'destinations string'
// has the correct amount, so we try to fetch it from that instead.
```

In my investigation "amount = 0" occurs when you send from account #1 to any address of account #1 (either main address or subaddress).

If I send from account #1 to account #2 (either main address or subaddress), the amount is correct.

xiphon replied:
Reproduced the issue. Looking into it.

xiphon replied:
https://github.com/monero-project/monero-gui/pull/3127 resolves the issue. Please confirm.

rating89us replied:
Yes, it's working now. Thanks

rating89us replied:
I won't fix this in this PR, since it will be fixed by your PR #3127 anyway

xiphon reviewed (commented):
`showTxDetails` changes look completely unrelated to the PR. if so, please submit unrelated changes in a separate PR.
